### PR TITLE
Refactor cluster condition handling

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

This PR refactors the cluster condition handling.

User-visible changes:

* We use a combination of the tag and the commit to show the kubermatic version to make the whole thing more readable
* We derive the kubermatic version from the vars in the resource package rather than the cluster status, to make sure its correct in the condition regardless of the question if the one controller that sets it in the status (cluster controller) did that already
* Instead of removing the condition when its false, we set its status to false
* We set the `LastTransitionTime` only when we can actually say for sure if it changed, rather than assuming absent -> present means it changed

Code changes:

* SetClusterCondition is exported to make it re-usable
* GetClusterCondition is exported to make it re-usable
* SetClusterCondition and GetClusterCondition are moved into their own package to avoid a recursive import of the `resources` package
* SetClusterCondition and GetClusterCondition got tests
* `SeedResourcesUpToDate` was renamed to `ClusterConditionSeedResourcesUpToDate` so ppl have a chance to find it via their IDEs completion
* `ClusterUpdateSuccesfulReason` was renamed to `ReasonClusterUpdateSuccessful` so ppl have a chance to find it via their IDE
* `SetSeedResourcesUpToDateCondition` retries cluster updates to avoid having it randomly fail due to concurrency issues
* The checks if SeedResources are up to date where factored out into their own function to allow evaluating them lazily instead of always evaluation everything and make it easier to follow the code

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

I am wondering if it wouldn't make more sense to have a small controller that sets the `SeedResourcesUpToDate` condition, because we currently have three controllers that do this but using the same underlying data which seems pretty pointless. What is more, they use their own reconciliation success to determine its state, so if one of them has an error it will fight with the others and the condition will keep flapping.

This will most likely entail having having a reconciliation success condition for each controller (which we need anyways, #3417) and use that plus the state of the resources to set the `SeedResourcesUpToDate` condition.

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

/assign @xrstf 
/cc @moadqassem

### CODE REVIEW

**Comment by alvaroaleman** [comment]
> I am wondering if it wouldn't make more sense to have a small controller that sets the SeedResourcesUpToDate condition, because we currently have three controllers that do this but using the same underlying data which seems pretty pointless. What is more, they use their own reconciliation success to determine its state, so if one of them has an error it will fight with the others and the condition will keep flapping.

Then again this implies that on Kubermatic upgrade, all clusters will be available for reconciliation which defeats the original purpose of that change. Maybe it makes more sense to simply ignore the `reconciliationSuccess` of the controllers for that?

File: api/pkg/controller/cluster/cluster_controller.go [hunk 1]
```diff
@@ -214,23 +215,25 @@ func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
...
 
-	if err := controllerutil.SetSeedResourcesUpToDateCondition(ctx, cluster, r, successfullyReconciled); err != nil {
+	if err := controllerutil.SetSeedResourcesUpToDateCondition(ctx, cluster, r.Client, successfullyReconciled); err != nil {
 		log.Errorw("failed to update clusters status conditions", zap.Error(err))
-		reconcileErr = fmt.Errorf("failed to set cluster status: %v after reconciliation was done with err=%v", err, reconcileErr)
+		errs = append(errs, err)
 	}
 
-	return *result, reconcileErr
+	return *result, utilerrors.NewAggregate(errs)
```
Comment (xrstf) [inline api/pkg/controller/cluster/cluster_controller.go:236]: This is a really nice refactoring :+1:

File: api/pkg/controller/util/util.go [hunk 2]
```diff
@@ -106,56 +108,84 @@ func ConcurrencyLimitReached(ctx context.Context, client ctrlruntimeclient.Clien
...
-					statefulSetHasUnfinishedUpdates = true
-				}
-			}
-		}
+	return updateCluster(ctx, client, cluster.Name, func(c *kubermaticv1.Cluster) {
+		kubermaticv1helper.SetClusterCondition(c,
+			kubermaticv1.ClusterConditionSeedResourcesUpToDate,
+			corev1.ConditionTrue,
+			kubermaticv1.ReasonClusterUpdateSuccessful,
+			"Some controlplane componends did not finish updating")
```
Comment (xrstf) [inline api/pkg/controller/util/util.go]: componenTs

File: api/pkg/controller/util/util.go [hunk 3]
```diff
@@ -106,56 +108,84 @@ func ConcurrencyLimitReached(ctx context.Context, client ctrlruntimeclient.Clien
...
+		//Get latest version
+		cluster := &kubermaticv1.Cluster{}
+		if err := client.Get(ctx, types.NamespacedName{Name: clusterName}, cluster); err != nil {
+			return err
+		}
+		// Apply modifications
+		modify(cluster)
+		// Update the cluster
+		return client.Update(ctx, cluster)
+	})
```
Comment (xrstf) [inline api/pkg/controller/util/util.go:190]: I've seen so many of these updateXXX functions lately that I wonder how they stand in relation to the reconciling system and if we should extract them somewhere. But then we would basically re-invent the codegen we already have for reconciling things.

Or am I overlooking something dramatically different between this and the reconciling package? Besides being a bit simpler to use, of course.

File: api/pkg/crd/kubermatic/v1/helper/helper_test.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+		},
+	}
+
+	for idx := range testCases {
+		tc := testCases[idx]
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			index, condition := GetClusterCondition(tc.cluster, tc.conditionType)
+			if index != tc.expectedIndex {
+				t.Errorf("expected index %d, got index %d", index, tc.expectedIndex)
```
Comment (xrstf) [inline api/pkg/crd/kubermatic/v1/helper/helper_test.go]: You accidentally flipped the two variables here.

File: api/pkg/controller/util/util.go (hunk 3 above)
Comment (alvaroaleman) [reply api/pkg/controller/util/util.go:190]: We do have a lot of them and yes, they all do the same. They are different from what is done in the reconciling package, as they are just about updating things, not upsert.

The main reason we have them so often is:
* We need them once per type as they are typed
* convenience

Generally, we could move all of them in typed factories but I thought this is easier to read and they are hard to get wrong anyways.

Comment (xrstf) [reply api/pkg/controller/util/util.go:190]: > they are hard to get wrong

Challenge accepted! :trollface:

**Pull request merged**